bundles/io/org.eclipse.smarthome.io.security/src/main/java/org/eclipse/smarthome/io/security/api/Cryptographic.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.io.security.api;
+
+import java.security.GeneralSecurityException;
+
+/**
+ * Implement this in {@link TlsNetworkServer} to get notifications of changed
+ * certificate contexts.
+ *
+ * @author David Graeff - Initial contribution
+ */
+public interface Cryptographic {
+    /**
+     * Encrypts the given text by the given password.
+     *
+     * @param password The password as byte array.
+     * @param plain The plain text to encrypt
+     * @return Returns the ciphered text.
+     * @throws GeneralSecurityException
+     */
+    char[] encrypt(byte[] password, char[] plain) throws GeneralSecurityException;
+
+    /**
+     * Decrypts the given text with the given password.
+     *
+     * @param password The password as byte array.
+     * @param cipher The cipher to decrypt
+     * @return Returns the plain text
+     * @throws GeneralSecurityException
+     */
+    char[] decrypt(byte[] password, char[] cipher) throws GeneralSecurityException;
+
+    /**
+     * Encrypts the given bytes by the given password.
+     *
+     * @param password The password as byte array.
+     * @param plain The plain bytes to encrypt
+     * @return Returns the ciphered text.
+     * @throws GeneralSecurityException
+     */
+    byte[] encrypt(byte[] password, byte[] plain) throws GeneralSecurityException;
+
+    /**
+     * Decrypts the given bytes with the given password.
+     *
+     * @param password The password as byte array.
+     * @param cipher The cipher to decrypt
+     * @return Returns the plain bytes
+     * @throws GeneralSecurityException
+     */
+    byte[] decrypt(byte[] password, byte[] cipher) throws GeneralSecurityException;
+
+    /**
+     * Returns the master password hash. Can be used to encrypt/decrypt, but
+     * remember that it will change, if the user changes the master password
+     * and old cipher will not be able to be decrypted anymore.
+     */
+    byte[] masterPasswordHash();
```

Comment:
I wanted to avoid such a listener. Everybody can register for this listener and an evil binding just logs the new plain text password and has full access. 

Instead I used a different pattern, see the security vault issue in the section Use-case.

Reply:
True, we definitely do not want to have a listener where everyone can obtain the password from :) But then we would need a "migration service". If you change the password I would not want to restart the system and wait for a re-encryption of all secrets. This should happen on the fly. If you offer your proposed `boolean migrateMasterPassword(String oldPassword, String newPassword)` method it should re-encrypt all secrets immediately and lock the configurations until this migration has finished and then store newly added data encrypted with the new password.

Reply:
Can we lock configadmin? A restart sounds more failsafe proof. It's a bad comparison, I know, but when I change the password of my router, I need to restart it. If I change a Windows login password, I need to logout and login again. I think people are used to the restart pattern if something central like the master password is changed. But open to suggestions.

Reply:
Another thought: What if you supply a new password but do **not** restart immediately but leave the system running for days/weeks... Which password will be used for new secrets that will be entered?

Also where do you store the new password, that you provided via rest api, in memory? But you do not have that memory if you restart the system...

Reply:
To stick to the router example: I get a hint message that the settings will only be applied if I restart the system.

The basic implementation does not have a rest service interface. The password can only be changed via environment variable on a restart.

An advanced (but not via ESH provided) implementation would have a PasswordChangeService or similar, which writes a new password to a file and forces a restart for example and offers a rest service interface. what do u think?

Reply:
I still do not quite understand how exactly this "environment variable" thing will look like. Do you have one variable "ESH_PASSWORD" which will be used by default and another one "ESH_PASSWORD_NEW" which the user has to set if he wants to change the current one? Who resets "ESH_PASSWORD_NEW" to "" once "ESH_PASSWORD" has been updated? And afterwards the user has to restart the whole ESH instance?

> PasswordChangeService or similar, which writes a new password to a file and forces a restart

How do you want to enforce a restart of the complete system from within a service? Does that work?

Also if the restart work, does the PasswordChangeService then start before all the other bindings, read the new password, migrate all secrets and then remove the file containing the new password, or what is your proposed work-flow?

Reply:
If a service has the access rights of an OSGi console, it can end and restart services. But that's not the point. 

We either have observers for a password change with old and new password and we agreed that's a bad idea, or we restart the framework. There's no other option that I can see here. 

My idea was a environment variable with the current password (/ new password) and if it is a change then there's a variable with the old password as well. We check the current password first and then the old password if existing to decrypt. If the old matches we re encrypt everything.

Reply:
Reencrypt. Github mobile doesn't allow edits. Sigh.

Reply:
Sorry I still don't get what you mean. :(

> My idea was a environment variable with the current password (/ new password)

Is that now one or two variables?

> and if it is a change

I am assuming "currentPW" and "newPW" are two variables.

> then there's a variable with the old password as well

So you copy the content of "currentPW" to "oldPW" and the content of "newPW" to "currentPW"?

> We check the current password first and then the old password

Why do you even introduce a third variable? Two are enough, aren't they? One for the old and one for the new password.

Reply:
Two variables only.

Reply:
It's current and old. Old is optional and should it be set for a change but it's not relevant if it still exists on later starts because we always first try current. And if that works we ignore old. But again: if there's another better approach I'm open to that. I only want the API here. And even the API can be extended, just not changed later on.